### PR TITLE
math: add initial Quaternion implementation

### PR DESCRIPTION
**Pull request by Olle-Lukowski**
Initial Quaternion implementation.

- [x] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

### CODE REVIEW

File: src/math/quat.zig [hunk 1]
```diff
@@ -15,9 +16,256 @@ pub fn Quat(comptime Scalar: type) type {
...
+            const zHalf = z * 0.5;
+
+            const sx = std.math.sin(xHalf);
+            const cx = std.math.cos(xHalf);
+            const sy = std.math.sin(yHalf);
+            const cy = std.math.cos(yHalf);
+            const sz = std.math.sin(zHalf);
+            const cz = std.math.cos(zHalf);
+
+            return init(sx * cy * cz + cx * sy * sz, cx * sy * cz - sx * cy * sz, cx * cy * sz + sx * sy * cz, cx * cy * cz - sx * sy * sz);
```
Comment (slimsag) [inline src/math/quat.zig:231]: Could we split the x, y, z, w components out into constants above this for slightly improved readability?

File: src/math/quat.zig [hunk 2]
```diff
@@ -15,9 +16,256 @@ pub fn Quat(comptime Scalar: type) type {
...
+        }
+
+        /// Returns the dot product of two quaternions.
+        pub inline fn dot(a: *const Quat(T), b: *const Quat(T)) T {
+            return a.v.x() * b.v.x() + a.v.y() * b.v.y() + a.v.z() * b.v.z() + a.v.w() * b.v.w();
+        }
+
+        /// Linearly interpolates between two quaternions.
+        pub inline fn lerp(a: *const Quat(T), b: *const Quat(T), t: T) Quat(T) {
+            return init(a.v.x() + t * (b.v.x() - a.v.x()), a.v.y() + t * (b.v.y() - a.v.y()), a.v.z() + t * (b.v.z() - a.v.z()), a.v.w() + t * (b.v.w() - a.v.w()));
```
Comment (slimsag) [inline src/math/quat.zig:241]: same here

File: src/math/quat.zig (hunk 1 above)
Comment (Olle-Lukowski) [reply src/math/quat.zig:231]: yeah, I assume you want the same in similar places?

File: src/math/quat.zig [hunk 3]
```diff
@@ -32,3 +280,212 @@ test "init" {
...
+
+test "normalize" {
+    const q = math.Quat.init(0.0, 0.0, 3.0, 4.0);
+    const expected = math.Quat.init(0.0, 0.0, 0.6, 0.8);
+    const actual = math.Quat.normalize(&q);
+
+    try testing.expect(f32, actual.v.x()).eqlApprox(expected.v.x(), 0.001);
+    try testing.expect(f32, actual.v.y()).eqlApprox(expected.v.y(), 0.001);
+    try testing.expect(f32, actual.v.z()).eqlApprox(expected.v.z(), 0.001);
+    try testing.expect(f32, actual.v.w()).eqlApprox(expected.v.w(), 0.001);
```
Comment (slimsag) [inline src/math/quat.zig:490]: Can we make all these tests just use `eql`? It should be approximate equality with the epsilon already, and so should be good enough for testing in general

File: src/math/quat.zig [hunk 4]
```diff
@@ -15,9 +16,266 @@ pub fn Quat(comptime Scalar: type) type {
         /// The underlying Vec type, e.g. math.Vec4, math.Vec4h, math.Vec4d
         pub const Vec = vec.Vec(4, Scalar);
 
-        pub inline fn init(x: T, y: T, z: T, w: T) Quat(Scalar) {
+        /// The Vec type used to represent axes, e.g. math.Vec3
+        pub const Axis = vec.Vec(3, Scalar);
```
Comment (slimsag) [inline src/math/quat.zig:20]: I'd suggest we rename this to `Vec3` instead - although `Axis` is obviously correct, a Quaternion axis is after all just a vector and I think it'll be less confusing to less math-inclined folks to see it's a `Vec3` than `Axis`

**Review by slimsag — APPROVED** [review]
Great work on this!

**Pull request merged**